### ISSUE TITLE
Continuous integration story for TLC model checking

### ISSUE THREAD

**Issue by lemmy** [tlaplus/tlaplus#306]
Today's development teams follow the practice of [continuous integration (CI)](https://en.wikipedia.org/wiki/Continuous_integration).  This trend is fueled by a plethora of (for-free) cloud services such as [TravisCI](https://travis-ci.org/), [Azure Pipelines](https://azure.microsoft.com/en-us/services/devops/pipelines/) or self-hosted automation servers such as [Jenkins](https://jenkins.io/). The advantage of CI - among others - is a central location where a record of the history of a software artifact is kept.  This history makes it e.g. easy to see the state of the software artifact. For TLC model checking (maybe for formal methods in general?) there exists no integration for the services mentioned above (the best thing appears to be a [home-grown docker solution](https://github.com/talex5/tla)).

Related issues are #157, #140, #225, #277

---------------------

In order to better support integrating TLC with CI services/systems, the following features should be implemented:

- [x] Time-bound model checking (end model after N minutes) 34cdc5b6ed6b3e708003b900bf1007d68a79d412
- [ ] Take checkpoint when time-bound model checking terminates (to allow teams to continue model checking from the checkpoint when necessary)
- [ ] Format TLC's "-coverage" output to be fed into commonly used coverage reporting frameworks (https://github.com/microsoft/azure-pipelines-tasks/tree/master/Tasks/Common/codecoverage-tools)
- [ ] Format TLC's error traces (assuming "-continue" enabled) for commonly used test reporting frameworks (formats such as [xUnit](https://github.com/xunit/xunit), [TAP](http://testanything.org/consumers.html), [junit](https://junit.org/))

**Comment by cipher1024** [comment]
> For TLC model checking (maybe for formal methods in general?) there exists no integration for [such] services

Little note: for interactive provers such as Lean, using Travis has been quite useful and it works right out of the box.

**Comment by lemmy** [comment]
@cipher1024 Thanks for note. Do you have a link?

**Comment by cipher1024** [comment]
Certainly: https://github.com/leanprover-community/mathlib/blob/master/.travis.yml

I must add though that model checking and interactive theorem proving, in CI, have the dual problem from each other. With interactive proving, we struggle to find a proof, we write a proof and once it's written it's quite easy to verify. We still run into performance issues but on the whole, the user takes care of the computationally intensive job of finding a proof. If the user breaks the build, the CI system will simply fail to approve a proof, even if the claim is still true.

With model checking, I think you need the CI system to redo the state exploration process which is very expensive. I don't know if you could have the user produce a certificate to say that they have done their homework and run the model checker on the latest version of their model.

**Comment by lemmy** [comment]
[PageQueue spec](https://github.com/lemmy/PageQueue) has a [CI setup](https://dev.azure.com/tlaplus/tlaplus/_build?definitionId=3&_a=summary)

**Comment by lemmy** [comment]
- [ ] Modify ```-workers``` parameter to take a value that indicates to use all available cores: ```-workers all```

**Comment by cipher1024** [comment]
Nice! When the CI fails, do you get a useful output?

**Comment by lemmy** [comment]
The recent build failures are all related to infrastructure problems which is why I cannot show this right now.  However, the CI system's output can be imported into the Toolbox to study the error trace if any. In addition, TLC can be instructed to write a checkpoint upon an invariant violation from which model checking can be restarted locally.  The work on [trace exploration](https://github.com/tlaplus/tlaplus/issues/393) might also be relevant here (@quaeler), because it will add functionality to TLC to export the error trace as valid TLA+.

**Comment by lemmy** [comment]
@ajdavis @will62794 @visualzhou Is there anything that you have found in your work that makes TLC difficult to use as part of CI?

**Comment by ajdavis** [comment]
No, I don't think so. We wrote scripts download-tlc.sh and model-check.sh to run on our CI servers, here:

https://github.com/mongodb/mongo/tree/master/src/mongo/db/repl/tla_plus

There have already been discussions about obtaining the command line arguments the TLA+ Toolbox passes to TLC, that's relevant to CI. Similarly for writing model configs: I start by configuring a model in the Toolbox and then adapt the .cfg file for CI. I suppose that workflow could be streamlined. Is there a way to view the text of a generated .cfg file in the TLA+ Toolbox? I had to search for the file in my directory tree.

**Comment by lemmy** [comment]
@ajdavis Do you see other use cases for showing the generated MC.cfg and MC.tla beyond this one?  Notice that the Toolbox shows the OS filesystem path of the model on the status line.  What might be more convenient, is to have support for opening the OS's file manager from the Toolbox's spec explorer.  That way, a user doesn't have to copy&paste the content of the `MC.cfg` and `MC.tla` out of the Toolbox.  Also, the issues https://github.com/tlaplus/tlaplus/issues/140 and https://github.com/tlaplus/tlaplus/issues/299 are related.

By the way, the actual TLC command-line gets logged to `~/.tlaplus/.metadata/.log` when TLC is launched.

**Comment by lemmy** [comment]
@ajdavis Looking at https://github.com/mongodb/mongo/blob/master/src/mongo/db/repl/tla_plus/model-check.sh, you may want to pass `-XX:MaxDirectMemorySize=1g -Dtlc2.tool.fp.FPSet.impl=tlc2.tool.fp.OffHeapDiskFPSet` to TLC.  `OffHeapDiskFPSet` is known to scale better.

Also, please consider adding `-Dutil.ExecutionStatisticsCollector.id=795f1f6c947011eab1d6b37484503bf1` to the command-line to have TLC report execution statistics that help us make it better.

**Comment by ajdavis** [comment]
"Do you see other use cases for showing the generated MC.cfg and MC.tla beyond this one?" Not right now, I think CI is the main use case for giving users convenient access to the MC files. A feature that opens their location in the file browser sounds great.

Thanks for the info about higher-performance flags?

**Comment by lemmy** [comment]
@ajdavis FYI https://github.com/tlaplus/tlaplus/commit/e6bd13ec971c1fa2a165b7068bc0091ab6c510cb

**Comment by lemmy** [comment]
> A feature that opens their location in the file browser sounds great.


![Peek 2020-05-15 13-54](https://user-images.githubusercontent.com/88777/82095411-d1b78880-96b3-11ea-9c46-b47357220c98.gif)

**Comment by JoshuaRowePhantom** [comment]
> 

@cipher1024 I second your approach. We don't actually need to run the model in a TLC session as part of CI. We just need proof that the model _has_ been run. A check of a model result file containing a hash of the input files would suffice.
